Add support for OpenRouter as a language model provider

imumesh18 opened the pull request:
This pull request adds full integration with OpenRouter, allowing users to access a wide variety of language models through a single API key.

**Implementation Details:**

* **Provider Registration:** Registers OpenRouter as a new language model provider within the application's model registry. This includes UI for API key authentication, token counting, streaming completions, and tool-call handling.
* **Dedicated Crate:** Adds a new `open_router` crate to manage interactions with the OpenRouter HTTP API, including model discovery and streaming helpers.
* **UI & Configuration:** Extends workspace manifests, the settings schema, icons, and default configurations to surface the OpenRouter provider and its settings within the UI.
* **Readability:** Reformats JSON arrays within the settings files for improved readability.

**Design Decisions & Discussion Points:**

* **Code Reuse:** I leveraged much of the existing logic from the `openai` provider integration due to the significant similarities between the OpenAI and OpenRouter API specifications.
* **Default Model:** I set the default model to `openrouter/auto`. This model automatically routes user prompts to the most suitable underlying model on OpenRouter, providing a convenient starting point.
* **Model Population Strategy:**
    * <strike>I've implemented dynamic population of available models by querying the OpenRouter API upon initialization.
    * Currently, this involves three separate API calls: one for all models, one for tool-use models, and one for models good at programming.
    * The data from the tool-use API call sets a `tool_use` flag for relevant models.
    * The data from the programming models API call is used to sort the list, prioritizing coding-focused models in the dropdown.</strike>
    * <strike>**Feedback Welcome:** I acknowledge this multi-call approach is API-intensive. I am open to feedback and alternative implementation suggestions if the team believes this can be optimized.</strike>
    * **Update: Now this has been simplified to one api call.**
* **UI/UX Considerations:**
    * <strike>Authentication Method: Currently, I've implemented the standard API key input in settings, similar to other providers like OpenAI/Anthropic. However, OpenRouter also supports OAuth 2.0 with PKCE. This could offer a potentially smoother, more integrated setup experience for users (e.g., clicking a button to authorize instead of copy-pasting a key). Should we prioritize implementing OAuth PKCE now, or perhaps add it as an alternative option later?</strike>(PKCE is not straight forward and complicated so skipping this for now. So that we can add the support and work on this later.)
    * <strike>To visually distinguish models better suited for programming, I've considered adding a marker (e.g., `</>` or `🧠`) next to their names. Thoughts on this proposal?</strike>. (This will require a changes and discussion across model provider. This doesn't fall under the scope of current PR).
    * OpenRouter offers 300+ models. The current implementation loads all of them. **Feedback Needed:** Should we refine this list or implement more sophisticated filtering/categorization for better usability?

**Motivation:**

This integration directly addresses one of the most highly upvoted feature requests/discussions within the Zed community. Adding OpenRouter support significantly expands the range of AI models accessible to users.

I welcome feedback from the Zed team on this implementation and the design choices made. I am eager to refine this feature and make it available to users.

ISSUES: https://github.com/zed-industries/zed/discussions/16576

Release Notes:

- Added support for OpenRouter as a language model provider.

imumesh18 committed:
Add OpenRouter language model provider

Introduces full integration with OpenRouter:
• Registers a new provider and models in the language‑model
  registry, including authentication UI, token counting,
  streaming completions and tool‑call handling.
• Adds dedicated `open_router` crate for HTTP API access,
  model discovery and streaming helpers.
• Extends workspace manifests, settings schema, icons and
  default config to surface the provider in the UI.
• Reformats JSON settings arrays for readability.

Broadens LLM choice without affecting existing providers.

Signed-off-by: Umesh Yadav <umesh4257@gmail.com>

imumesh18 commented:
The feature is working end to end. It is ready for review, i marked it draft so that I can do more extensive testing.

imumesh18 committed:
Prioritizes coding models in list

Sorts models alphabetically first.
Then moves models excelling at coding to the top for better visibility.

maxdeviant commented on assets/icons/ai_open_router.svg:
We’ll need to make sure this icon has the same scale and matches our existing icons.

imumesh18 replied:
Updated the image to be 16x16px like other icons of ai labs.

imumesh18 committed:
Replaces OpenRouter SVG with 16 px variant

Switches the icon to a 16×16 viewBox and simplified paths,
cutting file size and aligning with the standard UI icon set.

lj3954 commented on crates/open_router/src/open_router.rs:
```diff
@@ -0,0 +1,617 @@
+    }
+}
+
+#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
```
Since schemas are used for completions in settings.json, they need to represent available options (model IDs), not an entire struct. In these cases, schemas either have to be removed or implemented in a way that they have the app context

imumesh18 replied:
I actually also had the same doubt what to expose in the config and looked at https://github.com/zed-industries/zed/blob/6db974dd324e1b7f41f09a29c04e1234cc99723c/crates/lmstudio/src/lmstudio.rs#L44

Which implements the dynamic model and it had more than one property inside of it's Model Struct exposed. So just wanted to keep it consistent somewhat with that give token, description etc are required.

imumesh18 replied:
Also I don't think this actually works because couldn't found the code which handles model fetched at runtime for suggestion in the settings.

imumesh18 replied:
Removed schemar as it was not getting used anyway. Open for suggestion on how can we provide the modelIds where they are getting fetched at runtime instead of compile time.

lj3954 commented on crates/open_router/src/open_router.rs:
```diff
@@ -0,0 +1,617 @@
+    let mut models = fetch_models(client, api_url, None).await?;
+
+    // Get models that support tool calls
+    let tool_models = fetch_models(client, api_url, Some("supported_parameters=tools")).await?;
+    let tool_model_ids: Vec<String> = tool_models.iter().map(|m| m.id.clone()).collect();
+
+    // Get models that excel at coding
+    let coding_models = fetch_models(client, api_url, Some("category=programming")).await?;
+    let coding_model_ids: Vec<String> = coding_models.iter().map(|m| m.id.clone()).collect();
+
+    // Update model flags based on the specialized queries
+    for model in &mut models {
+        model.supports_tool_calls = tool_model_ids.contains(&model.id);
+        model.excels_at_coding = coding_model_ids.contains(&model.id);
+    }
```
These should probably be sets; using a vector for equality contains is just not that efficient. You also don't have to clone each string here. You can reference the IDs from the existing structs, making a HashSet<&str>. Or, since you're not using coding_models or tool_models again, you can just take ownership by replacing iter() with into_iter().

imumesh18 committed:
Adds OpenRouter to providers schema

Extends the allowed provider enumeration, enabling configs that
specify the "openrouter" backend to pass schema validation.

lj3954 commented on crates/open_router/src/open_router.rs:
```diff
@@ -0,0 +1,617 @@
+    !*v
+}
+
+#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
```
Are we sending this ModelEntry to the OpenRouter API? Why do we need a serialize implementation? I can only find reference to these structs being used to create a different `Model` struct.
I'm also not sure exactly the purpose of that. Serde will skip JSON keys which aren't in your struct during deserialization. You can remove fields which aren't used in Zed (e.g. Pricing, Architecture), and replace i64 with usize (OpenRouter's API spec claims context_length to be f64 anyways, presumably due to it being JS's number type; we can presume it will never be negative nor have decimals).

imumesh18 replied:
Fixed.

imumesh18 committed:
fix(open_router): optimize model fetching and handle context_length safely

imumesh18 committed:
fix clippy error

swrge commented:
As an avid user of both OpenRouter and Zed this is extremely important and exciting to me... Thanks you! Hope this gets merged 🙏🙏🙏

ChristianPerez34 commented:
Would it be possible to include extra headers for openrouter to display zed editor usage per model? 

Relevant discussion thread here: https://github.com/zed-industries/zed/discussions/28928

imumesh18 commented:
Hey @ChristianPerez34 , it's getting attached by default in this integration.

imumesh18 commented on crates/language_models/src/provider/open_router.rs:
```diff
@@ -0,0 +1,812 @@
+    model: &Model,
+    max_output_tokens: Option<u32>,
+) -> open_router::Request {
+    let stream = !model.id().starts_with("meta-llama/") && !model.id().starts_with("google/");
```
Thanks for spotting. Need to remove this as this only used for testing.

imumesh18 committed:
Remove unused stream variable in into_open_router

imumesh18 committed:
📎

imumesh18 committed:
code cleanup

imumesh18 committed:
Merge branch 'main' into openrouter

parsecgames commented:
please review and merge. ty

lj3954 commented:
Please don't provide such unproductive comments, it doesn't help anyone.

imumesh18 committed:
Fix tool call insertion and lower default temperature to 0.4

arrrrny commented:
@imumesh18 thats awesome. I have been using my own implementation and I can tell you from experience that 'tool support' filter is not always true. I did the same approach as you did and then I ended up testing out manually and weed out the models that dont support tools but tagged as supporting tools. Not that you can do anything but just giving you a heads up.

imumesh18 committed:
Merge branch 'main' into openrouter

arrrrny commented:
Also, I added a button to fetch models, since models can change, new models arrive. it would be a nice addition. 

```
                    h_flex()
                        .gap_2()
                        .child(
                            Button::new("fetch-models", "Fetch Models")
                                .label_size(LabelSize::Small)
                                .icon(Some(IconName::Play))
                                .icon_size(IconSize::Small)
                                .on_click(cx.listener(|this, _event, _window, cx| {
                                    this.retry_fetch_models(cx)
                                }))
                                .disabled(loading_models),
                        )
     ```

lj3954 commented:
> Also, I added a button to fetch models, since models can change, new models arrive. it would be a nice addition.
> 
> ```
>                     h_flex()
>                         .gap_2()
>                         .child(
>                             Button::new("fetch-models", "Fetch Models")
>                                 .label_size(LabelSize::Small)
>                                 .icon(Some(IconName::Play))
>                                 .icon_size(IconSize::Small)
>                                 .on_click(cx.listener(|this, _event, _window, cx| {
>                                     this.retry_fetch_models(cx)
>                                 }))
>                                 .disabled(loading_models),
>                         )
>      ```
> ```

I disagree. Generally you're not going to hear about the release of new models while you're already focused on a project. And if you are, it's likely a very small inconvenience to close and re-open the editor.

zikzak-wtf commented:
> > Also, I added a button to fetch models, since models can change, new models arrive. it would be a nice addition.
> > ```
> >                     h_flex()
> >                         .gap_2()
> >                         .child(
> >                             Button::new("fetch-models", "Fetch Models")
> >                                 .label_size(LabelSize::Small)
> >                                 .icon(Some(IconName::Play))
> >                                 .icon_size(IconSize::Small)
> >                                 .on_click(cx.listener(|this, _event, _window, cx| {
> >                                     this.retry_fetch_models(cx)
> >                                 }))
> >                                 .disabled(loading_models),
> >                         )
> >      ```
> > ```
> 
> I disagree. Generally you're not going to hear about the release of new models while you're already focused on a project. And if you are, it's likely a very small inconvenience to close and re-open the editor.

AS I check the current code, it only loads models when 'setttings' changed. Can you point what section of code it fetches models on Editor start? If thats the the case so be it. Also this fetch button can be useful with passing parameters. Some might just want 'free' models.

swrge commented:
> Also, I added a button to fetch models, since models can change, new models arrive. it would be a nice addition. 
> 
> ```
>                     h_flex()
>                         .gap_2()
>                         .child(
>                             Button::new("fetch-models", "Fetch Models")
>                                 .label_size(LabelSize::Small)
>                                 .icon(Some(IconName::Play))
>                                 .icon_size(IconSize::Small)
>                                 .on_click(cx.listener(|this, _event, _window, cx| {
>                                     this.retry_fetch_models(cx)
>                                 }))
>                                 .disabled(loading_models),
>                         )
>      ```

Imo this is the kind of polishing thing you do after the main feature is published and you have a fair amount of feedback from people using it..

arrrrny commented:
@imumesh18 Assistant is renamed to Agent. So this PR itself will never get merged as it. Will you adopt to the changes and open a new PR? I think your implementation is clean and decent. Otherwise I will open a PR with my implementation.

imumesh18 commented:
Thanks for letting me know. Will merge main and make the changes accordingly.

imumesh18 committed:
Merge remote-tracking branch 'origin/main' into openrouter

imumesh18 committed:
Add support for tool_choice in OpenRouter provider

imumesh18 committed:
📎

jipiboily commented:
I'm not a Rust dev, but let me know if you want someone to give it a shot before a review / merge before hand if that can help! 

Either way, thanks for your work and looking forward to seeing this shipped :)

gyzerok commented on crates/open_router/src/open_router.rs:
```diff
@@ -0,0 +1,535 @@
+
+    /// Indicates whether the model can handle OpenAI‑style tool/function calls.
+    #[serde(default, skip_serializing_if = "is_false")]
+    pub supports_tool_calls: bool,
```
The API for model listing returns a "supported_parameters" which is a list of strings. It looks like if model support tools then this list includes `"tool"` string and if it supports tool choice it includes `"tool_choice"`. Would it make sense to utilize this info instead of making a separate API call?

imumesh18 replied:
I agree even I don't like calling multiple api. I also haven't looked at the api response recently but here is the reference from conversation with openrouter team: https://github.com/zed-industries/zed/discussions/16576#discussioncomment-12957024. Not sure if they have added those parameters into the response but from this discussion thread during the time I made the changes they didn't have anyway for us to know whether or not we can use tool with this model. I might take a look at the response data again if these are added then probably we don't have to do multiple api calls and can do it one.

gyzerok replied:
[Here](https://openrouter.ai/docs/api-reference/list-available-models) is where I discovered it. You can also call the API right from the docs page to inspect the response. Hope that'll help

lj3954 replied:
> I agree even I don't like calling multiple api. I also haven't looked at the api response recently but here is the reference from conversation with openrouter team: [#16576 (reply in thread)](https://github.com/zed-industries/zed/discussions/16576#discussioncomment-12957024). Not sure if they have added those parameters into the response but from this discussion thread during the time I made the changes they didn't have anyway for us to know whether or not we can use tool with this model. I might take a look at the response data again if these are added then probably we don't have to do multiple api calls and can do it one.

Looks as though they have added a supported_parameters field. So you can use that rather than making another request.

imumesh18 replied:
Made the changes and closing this discussion

yeferson59 commented:
<img width="992" alt="Screenshot 2025-05-12 at 6 38 16 PM" src="https://github.com/user-attachments/assets/c7595916-126e-4fd1-8b11-b1517ccdc32f" />
I'm asking something in the chat and this error comes up, but since I don't know anything about rust and I get this

imumesh18 committed:
Update OpenRouter model handling and token counting

- Use "gpt-4o" encoding for token counting
- Always return None for max_output_tokens
- Add supported_parameters field to ModelEntry
- Refactor get_models to use supported_parameters for tool call support
- Simplify model fetching and flag updates

imumesh18 committed:
fix typo

imumesh18 committed:
Merge branch 'main' into openrouter

imumesh18 committed:
Refactor OpenRouter streaming event mapping and remove unused headers

- Move OpenRouter streaming event mapping logic into a new
  OpenRouterEventMapper struct for clarity and maintainability
- Remove http_referer and http_user_agent fields from OpenRouter request
  struct and related code, as these headers are now set directly in HTTP
  requests
- Update handling of tool result content to support images in the future

imumesh18 committed:
cleanup code

yeferson59 commented:
<img width="551" alt="Screenshot 2025-05-13 at 10 53 25 AM" src="https://github.com/user-attachments/assets/e605b476-80f4-446e-9c63-7ff60d4ec62a" />
@imumesh18 I get this warning when I'm testing OpenRouter support. Is this normal?
And by the way, it seems like too many models appear in the list. When I tried it, it works fine with chat, but when I answer, I get an error saying the window wasn't found.

yeferson59 commented:
> Hey @yeferson59 are you there on discord. We can move there, as every comment here pings 10+ people. Discord [windygod9986](https://discord.com/channels/@me/745557762737569805). You can help me debug on discord.

hey, I already sent you an invitation to your Discord with jeff2002

imumesh18 committed:
Merge branch 'main' into openrouter

parsecgames reviewed (approved):
Nice job!

imumesh18 committed:
cleanup code

parsecgames commented:
@imumesh18 would you be willing to setup artifact building in your fork and keep it updating against the upstream until it can get merged here? Seems like a few of us are interested.

imumesh18 committed:
Merge branch 'main' into openrouter

imumesh18 committed:
code cleanup

imumesh18 commented:
Hey @parsecgames , I don't think I will be able to put that much effort to setup the automated build given that it will require navigating significant portion of build config and it's not that straight forward. The best option is if someone from zed team adds a run-bundling tag to this PR every commit will be able to get it's own linux and macos build as part of the every ci action run which can help us achieve of getting a continuous build and people can download the artifact from latest action.

imumesh18 committed:
Merge branch 'main' into openrouter

imumesh18 committed:
Use full URL for HTTP-Referer header in requests

imumesh18 committed:
code cleanup

Michael-Obele commented:
Hi everyone, what's the update on this?

parsecgames commented:
> Hi everyone, what's the update on this?

It's fully functional. Just need to pull and build it yourself. The team will get  to it in due time. They likely have their hands full atm.

sysradium commented:
Yeah, it also looks absolutely fine to me. I would rather merge it and let it sit on main for some time so it everyone tests it. I tested it myself off the branch - seems to work to well.

cvanelteren commented:
Works great thanks!

cvanelteren commented:
Any hold up @maxdeviant ? Compiling zed takes a very long time and this seems good to go.

rstacruz commented:
This works pretty great!

One minor UX papercut I found is that OpenRouter model names can get very long, which leads to truncation:

![imagen](https://github.com/user-attachments/assets/7a09daf3-4cc0-4be6-a663-a83024529de6)

Not sure how to fix this, but I can live with it.

sysradium commented:
Maybe the full name can be put into a tooltip

parsecgames commented:
One way to handle it is to switch to a treelist with vendor as a parent. Other type of grouping can be done with a treelist. That being said, it will be a major change.

A simpler solution would be to show first x characters and last x characters rather than just first x character.

Another simple solution is multiline list item.

imumesh18 committed:
Merge branch 'main' into openrouter

imumesh18 committed:
Extract model name from display_name for UI clarity

imumesh18 committed:
📎

imumesh18 committed:
Merge branch 'main' into openrouter

imumesh18 committed:
📎

arrrrny commented:
@imumesh18  it does not generate Thread summary for me as it does no other providers. Clicking on generate dont work either.
Also sometimes, it just loose the token count and go down to 560 or less 1K from 30-40K
<img width="512" alt="Screenshot 2025-06-01 at 1 55 15 PM" src="https://github.com/user-attachments/assets/7bcc7073-ff09-4f74-95ff-06fb5f5ad66c" />

cvanelteren commented:
@imumesh18 I cannot confirm the issue raised by @arrrrny. For me it generates the summaries.

lj3954 commented:
Oh, looks like that's already more or less a thing: https://zed.dev/docs/ai/configuration#feature-specific-models (summary appears to be the title)

maxdeviant committed:
Merge branch 'main' into pr/29496

maxdeviant reviewed (approved):
Thank you!

The-Best-Codes commented:
Just saw that OpenRouter is supported in the latest stable Zed version. Thanks a lot everyone :heart:
